ntp-daemon/src/server.rs

```diff
@@ -0,0 +1,147 @@
+use std::{net::SocketAddr, sync::Arc};
+
+use ntp_proto::{
+    NtpAssociationMode, NtpClock, NtpHeader, NtpTimestamp, ReferenceId, SystemSnapshot,
+};
+use ntp_udp::UdpSocket;
+use tokio::{sync::RwLock, task::JoinHandle};
+use tracing::{instrument, warn};
+
+pub struct ServerTask<C: 'static + NtpClock + Send> {
+    socket: UdpSocket,
+    system: Arc<RwLock<SystemSnapshot>>,
+    clock: C,
+}
+
+impl<C: 'static + NtpClock + Send> ServerTask<C> {
+    pub fn spawn(
+        addr: SocketAddr,
+        system: Arc<RwLock<SystemSnapshot>>,
+        clock: C,
+        network_wait_period: std::time::Duration,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            let socket = loop {
+                match UdpSocket::new::<_, SocketAddr>(addr, None).await {
+                    Ok(socket) => break socket,
+                    Err(error) => {
+                        warn!(?error, "Could not open server socket");
+                        tokio::time::sleep(network_wait_period).await;
+                    }
+                }
+            };
+
+            // Unwrap should be safe because we know the socket was bound to a local addres just before
+            let _our_id = ReferenceId::from_ip(socket.as_ref().local_addr().unwrap().ip());
+
+            let mut process = ServerTask {
+                socket,
+                system,
+                clock,
+            };
+
+            process.serve().await
+        })
+    }
+
+    async fn generate_response(
+        &mut self,
+        input: NtpHeader,
+        recv_timestamp: NtpTimestamp,
+    ) -> NtpHeader {
+        let system = self.system.read().await;
+        let mut response = NtpHeader::new();
+        response.mode = NtpAssociationMode::Server;
+        response.stratum = system.stratum;
+        response.origin_timestamp = input.transmit_timestamp;
+        response.receive_timestamp = recv_timestamp;
+        response.reference_id = system.reference_id;
+        response.poll = input.poll;
+        response.precision = system.precision.log2();
+        response.root_delay = system.root_delay;
+        response.root_dispersion = system.root_dispersion;
+        response.transmit_timestamp = self.clock.now().expect("Failed to read time");
```

Comment:
why not use a struct initialization here? i.e. `NtpHeader { stratum: system.stratum, ..NtpHeader::new() }`

Reply:
Fixed